pyanaconda/modules/storage/iscsi/iscsi.py

```diff
@@ -0,0 +1,240 @@
+# Red Hat, Inc.
+#
+from blivet.iscsi import iscsi
+
+from pyanaconda.anaconda_loggers import get_module_logger
+from pyanaconda.core.signal import Signal
+from pyanaconda.dbus import DBus
+from pyanaconda.modules.common.base import KickstartBaseModule
+from pyanaconda.modules.common.constants.objects import ISCSI
+from pyanaconda.modules.storage.constants import IscsiInterfacesMode
+from pyanaconda.modules.storage.iscsi.discover import ISCSIDiscoverTask, ISCSILoginTask
+from pyanaconda.modules.storage.iscsi.iscsi_interface import ISCSIInterface, ISCSIDiscoverTaskInterface
+
+log = get_module_logger(__name__)
+
+
+class ISCSIModule(KickstartBaseModule):
+    """The iSCSI module."""
+
+    def __init__(self):
+        super().__init__()
+        self.reload_module()
+
+        self.initiator_changed = Signal()
+
+        self._iscsi_data = list()
+
+    def publish(self):
+        """Publish the module."""
+        DBus.publish_object(ISCSI.object_path, ISCSIInterface(self))
+
+    def reload_module(self):
+        """Reload the iscsi module."""
+        log.debug("Start up the iSCSI module.")
+        iscsi.startup()
+
+    @property
+    def initiator(self):
+        """The iSCSI initiator.
+
+        :return: a name of the initiator
+        """
+        return iscsi.initiator
+
+    def set_initiator(self, initiator):
+        """Set the iSCSI initiator.
+
+        :param initiator: a name of the initiator
+        """
+        if not iscsi.initiator_set:
+            iscsi.initiator = initiator
+            self.initiator_changed.emit()
+            log.debug("The iSCSI initiator is set to '%s'.", initiator)
+        else:
+            log.debug("The iSCSI initiator has already been set to '%s'.", iscsi.initiator)
+
+    def can_set_initiator(self):
+        """Can the initiator be set?
+
+        Once there there are active nodes logged in the initiator name can't be set.
+        """
+        return not iscsi.initiator_set
+
+    def get_interface_mode(self):
+        """Get the mode of interfaces used for iSCSI operations.
+
+        returns: an instance of IscsiInterfacesMode
+        """
+        mode = iscsi.mode
+        if mode == "none":
+            return IscsiInterfacesMode.UNSET
+        elif mode == "default":
+            return IscsiInterfacesMode.DEFAULT
+        elif mode == "bind":
+            return IscsiInterfacesMode.IFACENAME
+        else:
+            log.error("Unknown iSCSI interface mode %s set by blivet, using UNSET", mode)
+            return IscsiInterfacesMode.UNSET
+
+    def discover_with_task(self, target, credentials, interfaces_mode):
+        """Discover an iSCSI device.
+
+        :param target: the target information
+        :param credentials: the iSCSI credentials
+        :param interfaces_mode: required mode specified by IscsiInterfacesMode string value
+        :return: a DBus path to a task
+        """
+        task = ISCSIDiscoverTask(target, credentials, IscsiInterfacesMode(interfaces_mode))
```

Comment:
We should talk about this face to face.

The idea here is:
The mode is not set until any target was logged in using given mode. Until that, the discovery may request mode to be used (via intefaces_mode argument - DEFAULT or IFACENAME). It is also possible to check whether the mode is already set - GetInterfaceMode: UNSET -> the discovery can require any of the 2 modes, DEFAULT or IFACENAME -> the discovery has to use this mode.
This API is designed based on blivet/iscsi.py module and its mode attribute.

Other option could be ISCIDiscoverTask not having the interfaces_mode argument while the mode would be set for the module separately by module API
like
CanSetInterfaceMode()
SetIscsiInterfacesMode()
IscsiDiscoveryTask() .. mode taken from IscsiInterfacesMode

The former solution seems more user friendly and having better atomicity to me.

But we can discuss it later.

Reply:
Hmm, I think my request wasn't very clear. I just wanted to move the code `IscsiInterfacesMode(interfaces_mode)` from `discover_with_task` to `DiscoverWithTask`, so the module can always expect the enum instead of string.

Regarding your comment, yes, the first option somehow makes more sense, but shouldn't we handle the initiator the same way?

Reply:
Aaah, I see, moving here: https://github.com/rhinstaller/anaconda/pull/1960/commits/34acc3a6def0d601f92221405327a937c417e2d4
Yes, the initiator is using kind of the other type of API. I think I did it this way beacause ther is a case where initiator name can be set independently of any action (eg discovery): iscsiname kickstart command.
Also, I was lying in the CanSetInitiator docstring, the actual situation is different from mode setting, this: https://github.com/rhinstaller/anaconda/pull/1960/commits/3a6583d9f6b8534cd7cc549ba6971b4df4531304